man/mono.1
.\" 
.\" mono manual page.
.\" Copyright 2003 Ximian, Inc. 
.\" Copyright 2004-2011 Novell, Inc. 
.\" Copyright 2011-2012 Xamarin Inc
.\" Copyright 2013 7digital Media Ltd.
.\" Author:
.\"   Miguel de Icaza (miguel@gnu.org)
.\"
.TH Mono "mono"
.SH NAME
mono \- Mono's ECMA-CLI native code generator (Just-in-Time and Ahead-of-Time)
.SH SYNOPSIS
.PP
.B mono [options] file [arguments...]
.PP
.B mono-sgen [options] file [arguments...]
.SH DESCRIPTION
\fImono\fP is a runtime implementation of the ECMA Common Language
Infrastructure.  This can be used to run ECMA and .NET applications.
.PP
The runtime loads the specified
.I file
and optionally passes
the
.I arguments
to it.  The 
.I file
is an ECMA assembly.  They typically have a .exe or .dll extension.
.PP
These executables can reference additional functionality in the form
of assembly references.  By default those assembly references are
resolved as follows: the \fBmscorlib.dll\fR is resolved from the
system profile that is configured by Mono, and other assemblies are
loaded from the Global Assembly Cache (GAC).   
.PP
The runtime contains a native code generator that transforms the
Common Intermediate Language into native code.
.PP
The code generator can operate in two modes: Just-in-time compilation
(JIT) or Ahead-of-time compilation (AOT).  Since code can be
dynamically loaded, the runtime environment and the JIT are always
present, even if code is compiled ahead of time.
.PP
The runtime provides a number of configuration options for running
applications, for developing and debugging, and for testing and
debugging the runtime itself.
.PP
The \fImono\fP command uses the moving and generational SGen garbage collector
while the \fImono-boehm\fP command uses the conservative Boehm
garbage collector.
.SH PORTABILITY
On Unix-based systems, Mono provides a mechanism to emulate the 
Windows-style file access, this includes providing a case insensitive
view of the file system, directory separator mapping (from \\ to /) and
stripping the drive letters.
.PP
This functionality is enabled by setting the 
.B MONO_IOMAP 
environment variable to one of 
.B all, drive
and 
.B case.
.PP
See the description for 
.B MONO_IOMAP
in the environment variables section for more details.
.SH METHOD DESCRIPTIONS
A number of diagnostic command line options take as argument a method
description.   A method description is a textual representation that
can be used to uniquely identify a method.   The syntax is as follows:
.nf
[namespace]classname:methodname[(arguments)]
.fi
.PP
The values in brackets are optional, like the namespace and the
arguments.   The arguments themselves are either empty, or a
comma-separated list of arguments.   Both the
\fBclassname\fR
and
\fBmethodname\fR
can be set to the special value '*' to match any values (Unix shell
users should escape the argument to avoid the shell interpreting
this).
.PP
The arguments, if present should be a comma separated list of types
either a full typename, or for built-in types it should use the
low-level ILAsm type names for the built-in types,
like 'void', 'char', 'bool', 'byte', 'sbyte', 'uint16', 'int16', 'uint',
'int', 'ulong', 'long', 'uintptr', 'intptr', 'single', 'double', 'string', 'object'.
.PP
Pointer types should be the name of the type, followed by a '*',
arrays should be the typename followed by '[' one or more commas (to
indicate the rank of the array), and ']'.
.PP
Generic values should use '<', one or more type names, separated by
both a comma and a space and '>'.
.PP
By-reference arguments should include a "&" after the typename.
.PP
Examples:
.nf
*:ctor(int)		// All constructors that take an int as an argument
*:Main			// Methods named Main in any class
*:Main(string[])	// Methods named Main that take a string array in any class
.fi
.SH RUNTIME OPTIONS
The following options are available:
.TP
\fB--aot\fR, \fB--aot[=options]\fR
This option is used to precompile the CIL code in the specified
assembly to native code.  The generated code is stored in a file with
the extension .so.  This file will be automatically picked up by the
runtime when the assembly is executed.  
.Sp 
Ahead-of-Time compilation is most useful if you use it in combination
with the -O=all,-shared flag which enables all of the optimizations in
the code generator to be performed.  Some of those optimizations are
not practical for Just-in-Time compilation since they might be very
time consuming.
.Sp
Unlike the .NET Framework, Ahead-of-Time compilation will not generate
domain independent code: it generates the same code that the
Just-in-Time compiler would produce.   Since most applications use a
single domain, this is fine.   If you want to optimize the generated
code for use in multi-domain applications, consider using the
-O=shared flag.
.Sp
This pre-compiles the methods, but the original assembly is still
required to execute as this one contains the metadata and exception
information which is not available on the generated file.  When
precompiling code, you might want to compile with all optimizations
(-O=all).  Pre-compiled code is position independent code.
.Sp
Precompilation is just a mechanism to reduce startup time, increase
code sharing across multiple mono processes and avoid just-in-time
compilation program startup costs.  The original assembly must still
be present, as the metadata is contained there.
.Sp
AOT code typically can not be moved from one computer to another
(CPU-specific optimizations that are detected at runtime) so you
should not try to move the pre-generated assemblies or package the
pre-generated assemblies for deployment.    
.Sp
A few options are available as a parameter to the 
.B --aot 
command line option.   The options are separated by commas, and more
than one can be specified:
.RS
.ne 8
.TP
.I asmonly
Instructs the AOT compiler to output assembly code instead of an
object file.
.TP
.I bind-to-runtime-version
.Sp
If specified, forces the generated AOT files to be bound to the
runtime version of the compiling Mono.   This will prevent the AOT
files from being consumed by a different Mono runtime.
.TP
.I data-outfile=FILE.dll.aotdata
.Sp
This instructs the AOT code generator to output certain data
constructs into a separate file.   This can reduce the executable
images some five to twenty percent.   Developers need to then ship the
resulting aotdata as a resource and register a hook to load the data
on demand by using the 
.I mono_install_load_aot_data_hook
method.
.TP
.I direct-icalls
.Sp
When this option is specified, icalls (internal calls made from the
standard library into the mono runtime code) are invoked directly
instead of going through the operating system symbol lookup operation.
This requires use of the
.I static
option.
.TP
.I direct-pinvoke
.Sp
When this option is specified, P/Invoke methods are invoked directly
instead of going through the operating system symbol lookup operation.
This requires use of the
.I static
option.
.TP
.I dwarfdebug
Instructs the AOT compiler to emit DWARF debugging information. When
used together with the nodebug option, only DWARF debugging
information is emitted, but not the information that can be used at
runtime.
.TP
.I full
.Sp
This creates binaries which can be used with the --full-aot option.
.TP
.I hybrid
.Sp
This creates binaries which can be used with the --hybrid-aot option.
.TP
.I llvm
AOT will be performed with the LLVM backend instead of the Mono backend where possible. This will be slower to compile but most likely result in a performance improvement.
.TP
.I llvmonly
AOT will be performed with the LLVM backend exclusively and the Mono backend will not be used. The only output in this mode will be the bitcode file normally specified with the
.I llvm-outfile
option. Use of
.I llvmonly
automatically enables the
.I full
and
.I llvm
options. This feature is experimental.
.TP
.I llvmopts=[options]
Use this option to add more flags to the built-in set of flags passed to the
LLVM optimizer.   When you invoke the 
.I mono
command with the 
.I --aot=llvm 
it displays the current list of flags that are being passed to the 
.I opt
command.  
.I The list of possible flags that can be passed can be
obtained by calling the bundled 
.I opt 
program that comes with Mono, and calling it like this:
.nf

	opt --help

.fi

.TP
.I llvmllc=[options]
Use this option to add more flags to the built-in set of flags passed to the
LLVM static compiler (llc).   The list of possible flags that can be passed can be
obtained by calling the bundled 
.I llc
program that comes with Mono, and calling it like this:
.nf

	llc --help

.fi
.TP
.I mcpu=[native o generic]
cpu=native allows AOT mode to use all instructions current CPU supports, e.g. AVX2, SSE42, etc.
Default value is 'generic'.
.I mattr=[cpu feature]
Allows AOT code generator to use specified CPU features where possible including `System.Runtime.Intrinsics.*'.
E.g. `mattr=+avx2,mattr=-lzcnt' unlocks sse1-4.2, avx1-2 and disables lzcnt. 
It's useful for cross-compilation or when it's not possible to use `-mcpu=native' (which enables 
all cpu feature current cpu has).
.I llvm-outfile=[filename]
Gives the path for the temporary LLVM bitcode file created during AOT.
.I dedup
Each AOT module will typically contain the code for inflated methods and wrappers that
are called by code in that module. In dedup mode, we identify and skip compiling all of those
methods. When using this mode with fullaot, dedup-include is required or these methods will 
remain missing.
.TP
.I dedup-include=[filename]
In dedup-include mode, we are in the pass of compilation where we compile the methods
that we had previously skipped. All of them are emitted into the assembly that is passed
as this option. We consolidate the many duplicate skipped copies of the same method into one.
.TP
.TP
.I info
Print the architecture the AOT in this copy of Mono targets and quit.
.TP
.I interp
Generates all required wrappers, so that it is possible to run --interpreter without
any code generation at runtime.  This option only makes sense with \fBmscorlib.dll\fR.
Embedders can set
.TP
.I depfile=[filename]
Outputs a gcc -M style dependency file.

.nf
mono_jit_set_aot_mode (MONO_AOT_MODE_INTERP);
.fi
.ne
.TP
.I ld-flags
Additional flags to pass to the C linker (if the current AOT mode calls for invoking it).
.TP
.I llvm-path=<PREFIX>
Same for the llvm tools 'opt' and 'llc'.
.TP
.I msym-dir=<PATH>
Instructs the AOT compiler to generate offline sequence points .msym files.
The generated .msym files will be stored into a subfolder of <PATH> named as the
compilation AOTID.
.TP
.I mtriple=<TRIPLE>
Use the GNU style target triple <TRIPLE> to determine some code generation options, i.e.
--mtriple=armv7-linux-gnueabi will generate code that targets ARMv7. This is currently
only supported by the ARM backend. In LLVM mode, this triple is passed on to the LLVM
llc compiler.
.TP
.I nimt-trampolines=[number]
When compiling in full aot mode, the IMT trampolines must be precreated
in the AOT image.  You can add additional method trampolines with this argument.
Defaults to 512.
.TP
.I ngsharedvt-trampolines=[number]
When compiling in full aot mode, the value type generic sharing trampolines must be precreated
in the AOT image.  You can add additional method trampolines with this argument.
Defaults to 512.
.TP
.I nodebug
Instructs the AOT compiler to not output any debugging information.
.TP
.I no-direct-calls
This prevents the AOT compiler from generating a direct calls to a
method.   The AOT compiler usually generates direct calls for certain
methods that do not require going through the PLT (for example,
methods that are known to not require a hook like a static
constructor) or call into simple internal calls.
.TP
.I nrgctx-trampolines=[number]
When compiling in full aot mode, the generic sharing trampolines must be precreated
in the AOT image.  You can add additional method trampolines with this argument.
Defaults to 4096.
.TP
.I nrgctx-fetch-trampolines=[number]
When compiling in full aot mode, the generic sharing fetch trampolines must be precreated
in the AOT image.  You can add additional method trampolines with this argument.
Defaults to 128.
.TP
.I ntrampolines=[number]
When compiling in full aot mode, the method trampolines must be precreated
in the AOT image.  You can add additional method trampolines with this argument.
Defaults to 4096.
.TP
.I outfile=[filename]
Instructs the AOT compiler to save the output to the specified file.
.TP
.I print-skipped-methods
If the AOT compiler cannot compile a method for any reason, enabling this flag
will output the skipped methods to the console.
.TP
.I profile=[file]
Specify a file to use for profile-guided optimization. See the \fBAOT profiler\fR sub-section. To specify multiple files, include the
.I profile
option multiple times.
.TP
.I profile-only
AOT *only* the methods described in the files specified with the
.I profile
option. See the \fBAOT profiler\fR sub-section.
.TP
.I readonly-value=namespace.typename.fieldname=type/value
Override the value of a static readonly field. Usually, during JIT
compilation, the static constructor is ran eagerly, so the value of
a static readonly field is known at compilation time and the compiler
can do a number of optimizations based on it. During AOT, instead, the static
constructor can't be ran, so this option can be used to set the value of such
a field and enable the same set of optimizations.
Type can be any of i1, i2, i4 for integers of the respective sizes (in bytes).
Note that signed/unsigned numbers do not matter here, just the storage size.
This option can be specified multiple times and it doesn't prevent the static
constructor for the type defining the field to execute with the usual rules
at runtime (hence possibly computing a different value for the field).
.TP
.I save-temps,keep-temps
Instructs the AOT compiler to keep temporary files.
.TP
.I soft-debug
This instructs the compiler to generate sequence point checks that
allow Mono's soft debugger to debug applications even on systems where
it is not possible to set breakpoints or to single step (certain
hardware configurations like the cell phones and video gaming
consoles). 
.TP
.I static
Create an ELF object file (.o) or .s file which can be statically linked into an
executable when embedding the mono runtime. When this option is used, the object file
needs to be registered with the embedded runtime using the mono_aot_register_module
function which takes as its argument the mono_aot_module_<ASSEMBLY NAME>_info global
symbol from the object file:

.nf
extern void *mono_aot_module_hello_info;

mono_aot_register_module (mono_aot_module_hello_info);
.fi
.ne
.TP
.I stats
Print various stats collected during AOT compilation.
.TP
.I temp-path=[path]
Explicitly specify path to store temporary files created during AOT compilation.
.TP
.I threads=[number]
This is an experimental option for the AOT compiler to use multiple threads
when compiling the methods.
.TP
.I tool-prefix=<PREFIX>
Prepends <PREFIX> to the name of tools ran by the AOT compiler, i.e. 'as'/'ld'. For
example, --tool=prefix=arm-linux-gnueabi- will make the AOT compiler run
'arm-linux-gnueabi-as' instead of 'as'.
.TP
.I ld-name=NAME
One of the tools used for AOT builds is the linker. Its name differs between various
systems and it may happen that the assumed default name of the binary is not present.
If the toolchain used does not have a linker with the default name (e.g. Android NDK
r22 does not have the default 'ld' linker prefixed with 'tool-prefix' above, instead
it has prefixed 'ld.gold' and 'ld.bfd' linkers) this option can be used to set the
linker binary name. It will be prefixed with 'tool-prefix' to form the full linker
executable name.
.TP
.I verbose
Prints additional information about type loading failures.
.TP
.I write-symbols,no-write-symbols
Instructs the AOT compiler to emit (or not emit) debug symbol information.
.TP
.I no-opt
Instructs the AOT compiler tot no call opt when compiling with LLVM.
.PP
For more information about AOT, see: http://www.mono-project.com/docs/advanced/aot/
.RE
.TP
\fB--aot-path=PATH\fR
List of additional directories to search for AOT images.
.TP
\fB--apply-bindings=FILE\fR
Apply the assembly bindings from the specified configuration file when running
the AOT compiler.  This is useful when compiling an auxiliary assembly that is
referenced by a main assembly that provides a configuration file.  For example,
if app.exe uses lib.dll then in order to make the assembly bindings from
app.exe.config available when compiling lib.dll ahead of time, use:
.nf
	mono --apply-bindings=app.exe.config --aot lib.dll
.fi
.TP
\fB--assembly-loader=MODE\fR
If mode is \fBstrict\fR, Mono will check that the public key token, culture and version
of a candidate assembly matches the requested strong name.  If mode is \fBlegacy\fR, as
long as the name matches, the candidate will be allowed. \fBstrict\fR is the behavior
consistent with .NET Framework but may break some existing mono-based applications.
The default is \fBlegacy\fR.
.TP
\fB--attach=[options]\fR
Currently the only option supported by this command line argument is
\fBdisable\fR which disables the attach functionality.
.TP
\fB--config filename\fR
Load the specified configuration file instead of the default one(s).
The default files are /etc/mono/config and ~/.mono/config or the file
specified in the MONO_CONFIG environment variable, if set.  See the
mono-config(5) man page for details on the format of this file.
.TP
\fB--debugger-agent=[options]\fR 
This instructs the Mono runtime to
start a debugging agent inside the Mono runtime and connect it to a
client user interface will control the Mono process.
This option is typically used by IDEs, like the MonoDevelop or Visual Studio IDEs.
.RS
The configuration is specified using one of more of the following options:
.RS
.ne 8
.TP
.I address=host:port
.Sp
Use this option to specify the IP address where your debugger client is
listening to.
.TP
.I loglevel=LEVEL
.Sp
Specifies the diagnostics log level for 
.TP
.I logfile=filename
.Sp
Used to specify the file where the log will be stored, it defaults to
standard output.
.TP
.I server=[y/n]
Defaults to no, with the default option Mono will actively connect to the
host/port configured with the \fBaddress\fR option.  If you set it to 'y', it 
instructs the Mono runtime to start debugging in server mode, where Mono
actively waits for the debugger front end to connect to the Mono process.  
Mono will print out to stdout the IP address and port where it is listening.
.TP
.I setpgid=[y/n]
If set to yes, Mono will call \fBsetpgid(0, 0)\fR on startup, if that function
is available on the system. This is useful for ensuring that signals delivered
to a process that is executing the debuggee are not propagated to the debuggee,
e.g. when Ctrl-C sends \fBSIGINT\fR to the \fBsdb\fR tool.
.TP
.I suspend=[y/n]
Defaults to yes, with the default option Mono will suspend the vm on startup 
until it connects successfully to a debugger front end.  If you set it to 'n', in 
conjunction with \fBserver=y\fR, it instructs the Mono runtime to run as normal, 
while caching metadata to send to the debugger front end on connection..
.TP
.I transport=transport_name
.Sp
This is used to specify the transport that the debugger will use to
communicate.   It must be specified and currently requires this to
be 'dt_socket'. 
.ne
.RE
.TP
\fB--desktop\fR
Configures the virtual machine to be better suited for desktop
applications.  Currently this sets the GC system to avoid expanding
the heap as much as possible at the expense of slowing down garbage
collection a bit.
.TP
\fB--full-aot\fR
This flag instructs the Mono runtime to not
generate any code at runtime and depend exclusively on the code
generated from using mono --aot=full previously.  This is useful for
platforms that do not permit dynamic code generation, or if you need
to run assemblies that have been stripped of IL (for example using
mono-cil-strip).
.Sp
Notice that this feature will abort execution at runtime if a codepath
in your program, or Mono's class libraries attempts to generate code
dynamically.  You should test your software upfront and make sure that
you do not use any dynamic features.
.TP
\fB--full-aot-interp\fR
Same as --full-aot with fallback to the interpreter.
.TP
\fB--gc=boehm\fR, \fB--gc=sgen\fR
Selects the Garbage Collector engine for Mono to use, Boehm or SGen.
Currently this merely ensures that you are running either the
\fImono\fR or \fImono-sgen\fR commands.    This flag can be set in the
\fBMONO_ENV_OPTIONS\fR environment variable to force all of your child
processes to use one particular kind of GC with the Mono runtime.
.TP
\fB--gc-debug=[options]\fR
Command line equivalent of the \fBMONO_GC_DEBUG\fR environment variable.
.TP
\fB--gc-params=[options]\fR
Command line equivalent of the \fBMONO_GC_PARAMS\fR environment variable.
.TP
\fB--arch=32\fR, \fB--arch=64\fR
(Mac OS X only): Selects the bitness of the Mono binary used, if
available. If the binary used is already for the selected bitness, nothing
changes. If not, the execution switches to a binary with the selected
bitness suffix installed side by side (for example, '/bin/mono --arch=64'
will switch to '/bin/mono64' iff '/bin/mono' is a 32-bit build).
.TP
\fB--help\fR, \fB-h\fR
Displays usage instructions.
.TP
\fB--interpreter\fR
The Mono runtime will use its interpreter to execute a given assembly.
The interpreter is usually slower than the JIT, but it can be useful on
platforms where code generation at runtime is not allowed.
.TP
\fB--hybrid-aot\fR
This flag allows the Mono runtime to run assemblies
that have been stripped of IL, for example using mono-cil-strip. For this to
work, the assembly must have been AOT compiled with --aot=hybrid.

This flag is similar to --full-aot, but it does not disable the JIT. This means
you can use dynamic features such as System.Reflection.Emit.
.TP
\fB--llvm\fR
If the Mono runtime has been compiled with LLVM support (not available
in all configurations), Mono will use the LLVM optimization and code
generation engine to JIT or AOT compile.     
.Sp
For more information, consult: http://www.mono-project.com/docs/advanced/mono-llvm/
.TP
\fB--ffast-math\fR
This flag allows Mono and LLVM to apply aggressive floating point optimizations.
Can break IEEE754 compliance.
.TP
\fB--nollvm\fR
When using a Mono that has been compiled with LLVM support, it forces
Mono to fallback to its JIT engine and not use the LLVM backend.
.TP
\fB--optimize=MODE\fR, \fB-O=MODE\fR
MODE is a comma separated list of optimizations.  They also allow
optimizations to be turned off by prefixing the optimization name with
a minus sign.
.Sp
In general, Mono has been tuned to use the default set of flags,
before using these flags for a deployment setting, you might want to
actually measure the benefits of using them.    
.Sp
The following optimization flags are implemented in the core engine:
.nf
             abcrem     Array bound checks removal
             all        Turn on all optimizations
             aot        Usage of Ahead Of Time compiled code
             branch     Branch optimizations
             cfold      Constant folding
             cmov       Conditional moves [arch-dependency]
             deadce     Dead code elimination
             consprop   Constant propagation
             copyprop   Copy propagation
             fcmov      Fast x86 FP compares [arch-dependency]
             float32	Perform 32-bit float arithmetic using 32-bit operations
             gshared    Enable generic code sharing.
             inline     Inline method calls
             intrins    Intrinsic method implementations
             linears    Linear scan global reg allocation
             leaf       Leaf procedures optimizations
             loop       Loop related optimizations
             peephole   Peephole postpass
             precomp    Precompile all methods before executing Main
             sched      Instruction scheduling
             shared     Emit per-domain code
             sse2       SSE2 instructions on x86 [arch-dependency]
             tailc      Tail recursion and tail calls
.fi
.Sp
For example, to enable all the optimization but dead code
elimination and inlining, you can use:
.nf
	-O=all,-deadce,-inline
.fi
.Sp
The flags that are flagged with [arch-dependency] indicate that the
given option if used in combination with Ahead of Time compilation
(--aot flag) would produce pre-compiled code that will depend on the
current CPU and might not be safely moved to another computer. 
.RS
.ne 8
.PP
The following optimizations are supported
.TP
.I float32
Requests that the runtime performn 32-bit floating point operations
using only 32-bits.   By default the Mono runtime tries to use the
highest precision available for floating point operations, but while
this might render better results, the code might run slower.   This
options also affects the code generated by the LLVM backend.
.TP
.I inline
Controls whether the runtime should attempt to inline (the default),
or not inline methods invocations
.ne
.RE
\fB--response=FILE\fR
Provides a response file, this instructs the Mono command to read
other command line options from the specified file, as if the
options had been specified on the command line.   Useful when you have
very long command lines.
.TP
\fB--runtime=VERSION\fR
Mono supports different runtime versions. The version used depends on the program
that is being run or on its configuration file (named program.exe.config). This option
can be used to override such autodetection, by forcing a different runtime version
to be used. Note that this should only be used to select a later compatible runtime
version than the one the program was compiled against. A typical usage is for
running a 1.1 program on a 2.0 version:
.nf
         mono --runtime=v2.0.50727 program.exe
.fi
.TP
\fB--server\fR
Configures the virtual machine to be better suited for server

NEWS
This is a useless commit

Release notes for Mono are hosted on the web, please see:

	http://www.mono-project.com/docs/about-mono/releases/


premake/mono.lua
MONO_RUNTIME_ROOT = MONO_ROOT .. "mono/"

function SetupSGen()
	local f = filter()

	defines
	{
		"HAVE_SGEN_GC",
		"HAVE_MOVING_COLLECTOR",
		"HAVE_WRITE_BARRIERS",
		"MONO_DLL_EXPORT"
	}

	files
	{
		MONO_RUNTIME_ROOT .. "metadata/sgen-*.*"
	}

	filter "system:windows"
		files { MONO_RUNTIME_ROOT .. "metadata/sgen-os-win*.c" }

	filter "system:not windows"
		files { MONO_RUNTIME_ROOT .. "metadata/sgen-os-posix.c" }

	filter "system:macosx"
		files { MONO_RUNTIME_ROOT .. "metadata/sgen-os-mach.c" }

	filter(f)
end

function SetupConfigDefines()
	defines
	{
		"HAVE_DECL_INTERLOCKEDCOMPAREEXCHANGE64=1",
		"HAVE_DECL_INTERLOCKEDEXCHANGE64=1",
		"HAVE_DECL_INTERLOCKEDINCREMENT64=1",
		"HAVE_DECL_INTERLOCKEDDECREMENT64=1",
		"HAVE_DECL_INTERLOCKEDADD=1",
		"HAVE_DECL_INTERLOCKEDADD64=1",
		"HAVE_DECL_INTERLOCKEDCOMPAREEXCHANGE64=1",
		"HAVE_DECL___READFSDWORD=1",
	}
end

function GenerateConfig()
	if not os.is("windows") then
		return
	end

	print('Generating Windows config.h')
	os.copyfile(MONO_ROOT .. "winconfig.h", gendir .. "/config.h")
end

function GenerateVersion()
	print('Generating version.h')

	local contents
	if os.isdir(MONO_ROOT .. ".git") then
		local branches = os.outputof("set LANG=C && git branch")
		local branch = string.gmatch(branches, "^\* (%w+)")()
		local version = os.outputof("set LANG=C && git log --no-color --first-parent -n1 --pretty=format:%h")
		contents = string.format("#define FULL_VERSION \"%s/%s\"", branch, version)
	else
		contents = "#define FULL_VERSION \"tarball\""
	end

	file = io.open(path.getabsolute(gendir .. "/version.h"), "w+")
	file:write(contents)
	file:close()
end

function SetupMonoIncludes(monoRuntimeRoot)
	includedirs
	{
		gendir,
		monoRuntimeRoot,
		monoRuntimeRoot .. "..",
		monoRuntimeRoot .. "../eglib/src",
		monoRuntimeRoot .. "utils/",
	}
	print(monoRuntimeRoot .. "../eglib/src")
end

function SetupMonoLinks()
	local f = filter()

	links
	{
		"eglib",
		"libmonoruntime",
		"libmonoutils"
	}

	filter "system:windows"

		links
		{
			"Mswsock",
			"ws2_32",
			"psapi",
			"version",
			"winmm",
		}

	filter()
end

function GenerateMachineDescription(arch)
	local prj = premake.api.scope.project.location
	local abs = path.getabsolute(MONO_RUNTIME_ROOT .. 'mini/cpu-' .. arch .. '.md')
	local input = path.getrelative(prj, abs)	
	local out = gendir .. '/cpu-' .. arch .. '.h'
	local desc = arch .. '_desc'

	prebuildcommands
	{
		'"%{cfg.targetdir}/genmdesc" ' .. out .. ' ' .. desc .. ' ' .. input
	}
end

GenerateConfig()
GenerateVersion()

project "mono"
	
	kind "ConsoleApp"
	language "C"

	files
	{
		MONO_RUNTIME_ROOT .. "mini/main.c",
	}

	includedirs
	{
		gendir,
		MONO_RUNTIME_ROOT .. "../",
		MONO_RUNTIME_ROOT .. "../eglib/src/"
	}	
	
	links
	{
		"eglib",
		"libmono",
		"libmonoruntime",
		"libmonoutils"
	}

	filter "action:vs*"
		defines { "_CRT_SECURE_NO_WARNINGS", "_CRT_NONSTDC_NO_DEPRECATE" }
		SetupConfigDefines()


project "libmono"

	kind "SharedLib"
	language "C"

	defines
	{
		"__default_codegen__",
		"HAVE_CONFIG_H",
		"MONO_DLL_EXPORT"
	}
	
	SetupMonoIncludes(MONO_RUNTIME_ROOT)
	SetupMonoLinks()

	files
	{
		MONO_RUNTIME_ROOT .. "mini/*.c",
		MONO_RUNTIME_ROOT .. "mini/*.h",
	}

	excludes
	{
		-- Archicture-specific files
		MONO_RUNTIME_ROOT .. "mini/*-alpha.*",
		MONO_RUNTIME_ROOT .. "mini/*-amd64.*",
		MONO_RUNTIME_ROOT .. "mini/*-arm.*",
		MONO_RUNTIME_ROOT .. "mini/*-arm64.*",
		MONO_RUNTIME_ROOT .. "mini/*-hppa.*",
		MONO_RUNTIME_ROOT .. "mini/*-ia64.*",
		MONO_RUNTIME_ROOT .. "mini/*-llvm.*",
		MONO_RUNTIME_ROOT .. "mini/*-mips.*",
		MONO_RUNTIME_ROOT .. "mini/*-ppc.*",
		MONO_RUNTIME_ROOT .. "mini/*-s390*.*",
		MONO_RUNTIME_ROOT .. "mini/*-sparc.*",
		MONO_RUNTIME_ROOT .. "mini/*-x86.*",

		-- Platform-specific files
		MONO_RUNTIME_ROOT .. "mini/*-windows.*",
		MONO_RUNTIME_ROOT .. "mini/*-darwin.*",
		MONO_RUNTIME_ROOT .. "mini/*-posix.*",

		-- Tools
		MONO_RUNTIME_ROOT .. "mini/fsacheck.c",
		MONO_RUNTIME_ROOT .. "mini/genmdesc.c",
		MONO_RUNTIME_ROOT .. "mini/main.c",
	}

	dependson { "genmdesc" }
	
	filter "platforms:x32"
		GenerateMachineDescription('x86')
		files
		{
			MONO_RUNTIME_ROOT .. "mini/*-x86.c",
		}

	filter "platforms:x64"
		GenerateMachineDescription('amd64')
		files
		{
			MONO_RUNTIME_ROOT .. "mini/*-amd64.c",
		}	

	filter "system:windows"
		SetupWindowsDefines()
		files
		{
			MONO_RUNTIME_ROOT .. "mini/*-windows.c"
		}

	filter "system:not windows"
		files
		{
			MONO_RUNTIME_ROOT .. "mini/*-posix.c"
		}

	filter "system:macosx"
		files
		{
			MONO_RUNTIME_ROOT .. "mini/*-darwin.c"
		}

	filter "action:vs*"
		defines { "_CRT_SECURE_NO_WARNINGS", "_CRT_NONSTDC_NO_DEPRECATE" }
		SetupConfigDefines()
		SetupMSVCWarnings()
		buildoptions
		{
			"/wd4018", -- signed/unsigned mismatch
			"/wd4244", -- conversion from 'x' to 'y', possible loss of data
			"/wd4133", -- incompatible types - from 'x *' to 'y *'
			"/wd4715", -- not all control paths return a value
			"/wd4047", -- 'x' differs in levels of indirection from 'y'
		}
		linkoptions
		{
			"/ignore:4049", -- locally defined symbol imported
			"/ignore:4217", -- locally defined symbol imported in function
		}

project "libmonoruntime"

	kind "StaticLib"
	language "C"
	
	defines
	{
		"HAVE_CONFIG_H",
	}

	SetupMonoIncludes(MONO_RUNTIME_ROOT)
	
	files
	{
		MONO_RUNTIME_ROOT .. "metadata/*.c",
		MONO_RUNTIME_ROOT .. "metadata/*.h",
		MONO_RUNTIME_ROOT .. "sgen/*.c",
		MONO_RUNTIME_ROOT .. "sgen/*.h",
	}

	excludes
	{
		MONO_RUNTIME_ROOT .. "metadata/threadpool-ms-io-*",

		-- GC-specific files
		MONO_RUNTIME_ROOT .. "metadata/boehm-gc.c",
		MONO_RUNTIME_ROOT .. "metadata/null-gc.c",
		MONO_RUNTIME_ROOT .. "metadata/sgen-*.*",
		MONO_RUNTIME_ROOT .. "metadata/sgen-os-*.*",
		
		-- Platform-specific files
		MONO_RUNTIME_ROOT .. "metadata/console-unix.c",
		MONO_RUNTIME_ROOT .. "metadata/console-win32.c",
		MONO_RUNTIME_ROOT .. "metadata/coree.c",

		-- Tools
		MONO_RUNTIME_ROOT .. "metadata/monodiet.c",
		MONO_RUNTIME_ROOT .. "metadata/monosn.c",
		MONO_RUNTIME_ROOT .. "metadata/pedump.c",
		MONO_RUNTIME_ROOT .. "metadata/tpool-*.c",
		MONO_RUNTIME_ROOT .. "metadata/test-*.c",
	}

	SetupSGen()

	filter "system:windows"
		SetupWindowsDefines()
		defines { "_WINSOCK_DEPRECATED_NO_WARNINGS" }
		files
		{
			MONO_RUNTIME_ROOT .. "metadata/coree.c",
		}

	filter "system:linux"
		files
		{
			MONO_RUNTIME_ROOT .. "metadata/tpool-epoll.c"
		}
		
	filter "system:macosx or freebsd"
		files
		{
			MONO_RUNTIME_ROOT .. "metadata/tpool-kqueue.c"
		}

	filter "action:vs*"
		defines { "_CRT_SECURE_NO_WARNINGS", "_CRT_NONSTDC_NO_DEPRECATE" }
		SetupConfigDefines()
		SetupMSVCWarnings()


project "libmonoutils"

	kind "StaticLib"
	language "C"
	
	defines
	{
		"HAVE_CONFIG_H",
	}

	SetupMonoIncludes(MONO_RUNTIME_ROOT)
	
	files
	{
		MONO_RUNTIME_ROOT .. "utils/*.c",
		MONO_RUNTIME_ROOT .. "utils/*.h",
	}

	excludes
	{
		MONO_RUNTIME_ROOT .. "utils/sha1.c",

		-- Platform-specific files
		MONO_RUNTIME_ROOT .. "utils/atomic.c",
		MONO_RUNTIME_ROOT .. "utils/mono-hwcap-*.*",
		MONO_RUNTIME_ROOT .. "utils/mono-threads-*.c",

		-- Tools
		MONO_RUNTIME_ROOT .. "utils/mono-embed.c",
	}

	files { MONO_RUNTIME_ROOT .. "utils/mono-threads-state-machine.c" }

	filter "architecture:arm*"
		files { MONO_RUNTIME_ROOT .. "utils/mono-hwcap-arm.*" }

	filter "architecture:x32 or x64"
		files { MONO_RUNTIME_ROOT .. "utils/mono-hwcap-x86.*" }

	filter "system:windows"
		SetupWindowsDefines()
		files
		{
			MONO_RUNTIME_ROOT .. "metadata/coree.c",
			MONO_RUNTIME_ROOT .. "utils/mono-threads-windows.c",
		}
		links
		{
			"Mswsock",
			"ws2_32",
			"psapi",
			"version",
			"winmm",
			"eglib",
		}

	filter "system:not windows"
		files
		{
			MONO_RUNTIME_ROOT .. "utils/mono-threads-posix.c",
		}

	filter "system:macosx"
		files
		{
			MONO_RUNTIME_ROOT .. "utils/mono-threads-mach.c",
		}

	filter "action:vs*"
		defines { "_CRT_SECURE_NO_WARNINGS", "_CRT_NONSTDC_NO_DEPRECATE" }
		SetupConfigDefines()
		SetupMSVCWarnings()
		buildoptions { "/wd4273", "/wd4197" }

